File: Assignment_2/C/Main.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include <omp.h>

#include "ObjectiveFunctions.h"
#include "UtilFunctions.h"

double** initInitialPopulation(int populationSize, int chromosomeSize, double lowerBound, double upperBound);

void evolution(double **initialPopulation, int generations, int populationSize, int chromosomeSize, int recombCoefM, int nrSelectedParents, int selectionStrat, int topSurvivors);
double** nextGeneration (double **initialPopulation, int generations, int populationSize, int chromosomeSize, int recombCoefM);

double* populationFitness(double** individuals, int populationSize, int chromosomeSize, int obectiveFunction);
double individualFitness(double* individual, int chromosomeSize, int obectiveFunction);

int main(int argc, char *argv[]){

    int populationSize = atoi(argv[1]), chromosomeSize = atoi(argv[2]), generations = atoi(argv[3]);
    double lowerBound = atof(argv[4]), upperBound = atof(argv[5]);
    int recombCoefM = atoi(argv[6]), nrSelectedParents = atoi(argv[7]), selectionStrat = atoi(argv[7]), topSurvivors = atoi(argv[8]);

    srand(time(NULL));

    printf("Population size = %d\nChromosome size (Number of genes)) = %d\nGenerations = %d\n", populationSize, chromosomeSize, generations);
    printf("Lowerbound = %lf\nUpperbound = %lf\n", lowerBound, upperBound);
    printf("Recombination coefficient method = %d\nNumberParents Selected for each offspring = %d\nSelection Strategy = %d\n", recombCoefM, nrSelectedParents, selectionStrat);
    printf("Number of top survivors (from collective parent offspring population) = %d\n\n", topSurvivors);

    // Check if parameters are valid


    // Generate Random initial population
    double **initialPopulation = initInitialPopulation(populationSize, chromosomeSize, lowerBound, upperBound);
    printDouble2DArrayPtr(initialPopulation, populationSize, chromosomeSize);

    double *fitness = populationFitness(initialPopulation, populationSize, chromosomeSize, 0);
    printDoubleArray(fitness, populationSize);

    Evolution for number of generations
    evolution(individuals, generations, populationSize, chromosomeSize, recombCoefM, nrSelectedParents, selectionStrat, topSurvivors);

    freeDouble2DArray(initialPopulation, populationSize);
    free(fitness);

    return 0;
}

double** initInitialPopulation(int populationSize, int chromosomeSize, double lowerBound, double upperBound) {

    double **individuals = (double**) malloc (populationSize * sizeof(double*));
 
    for (int i = 0; i < populationSize; i++) {
        *(individuals + i) = (double*) malloc(chromosomeSize * sizeof(double));
    }

    for (int i = 0; i < populationSize; i++) {
        for (int j = 0; j < chromosomeSize; j++) {
            *(*(individuals + i) + j) = rand_range_double(lowerBound, upperBound); // OR individuals[i][j] = ...
        }
    }

    return individuals;
}

void evolution(double **initialPopulation, int generations, int populationSize, int chromosomeSize, int recombCoefM, int nrSelectedParents, int selectionStrat, int topSurvivors) {

    if (selectionStrat == 0) {
        //Random parent selection
    } else if (selection == 1) {
        //Tournament selection
    } else if () {

    }s


    double **parents = copy_array(initialPopulation, populationSize, chromosomeSize);

    for (int g = 0; g < generations; g++) {





        double **offspring = nextGeneration(parents, generations, populationSize, chromosomeSize, recombCoefM);


    }

    freeDouble2DArray(parents, populationSize);

}

double** nextGeneration (double **parents, int generations, int populationSize, int chromosomeSize, int recombCoefM) {


    return NULL;

}

double* populationFitness(double** individuals, int populationSize, int chromosomeSize, int obectiveFunction) {
    
    double *fitnessArray = (double*) malloc(populationSize * sizeof(double));

    #pragma omp parallel for 
    for (int i = 0; i < populationSize; i++) {
        *(fitnessArray + i) = individualFitness(*(individuals + i), chromosomeSize, obectiveFunction);

        //int tid = omp_get_thread_num();
        //printf("The thread %d  executes i = %d\n", tid, i);
    }
    return fitnessArray;
}

double individualFitness(double* individual, int chromosomeSize, int obectiveFunction) {

    if (obectiveFunction == 0) {
        return absolute(individual, chromosomeSize);
    }

}

File: Assignment_2/Python/Main.py
import sys
import Evolution as ev
import threading
import time

population_size = int(sys.argv[1])
chromosome_size = int(sys.argv[2])
generations = int(sys.argv[3])

nr_breeding_parents = int(sys.argv[4])
nr_offspring = int(sys.argv[5])
nr_tournament = int(sys.argv[6])

parent_selection_m = int(sys.argv[7])
recomb_coef_m = int(sys.argv[8])

objective_function = int(sys.argv[9])
#lower_bound = float(sys.argv[10])
#upper_bound = float(sys.argv[11])

print(f"{'Population Size:' : <30}{population_size}")
print(f"{'Chromosome Size:' : <30}{chromosome_size}")
print(f"{'Generations:' : <30}{generations}")
print(f"{'Nr of Breeding Parents:' : <30}{nr_breeding_parents}")
print(f"{'Nr of Offspring:' : <30}{nr_offspring}")
print(f"{'Nr Parents in Tournament:' : <30}{nr_tournament}")
print(f"{'Parent Selection Method:' : <30}{parent_selection_m}")
print(f"{'Recomb Coef Method:' : <30}{recomb_coef_m}")
print(f"{'Obective Function:' : <30}{objective_function}")
#print(f"{'Lower Bound:' : <30}{lower_bound}")
#print(f"{'Upper Bound:' : <30}{upper_bound}")
print()

objective_function_names = ["Absolute", "Ackley 1", "Alpine 1", "Step 2", "Schwefel 2.23", "Step 3", "Shubert 4", "Discus", "Egg Crate", "Deb 1"]
lower_bounds = [-100.0, -32.0, -10.0, -100.0, -10.0, -5.12, -10.0, -100.0, -5.0, -1.0]
upper_bounds = [100.0, 32.0, 10.0, 100.0, 10.0, 5.12, 10.0, 100.0, 5.0, 1.0]

if (objective_function == -1): # Go over all functions

    threads = []

    for f in range(10):

        lower_bound = lower_bounds[f]
        upper_bound = upper_bounds[f]

        objective_function_name = objective_function_names[f]
        t = threading.Thread(target = ev.evolution, args = [population_size, chromosome_size, generations, nr_breeding_parents, nr_offspring,
                                                            nr_tournament, parent_selection_m, recomb_coef_m, f, objective_function_name,
                                                            lower_bound, upper_bound])
        t.start()
        threads.append(t)
        
    for t in threads:
        t.join()
    
else:
    lower_bound = lower_bounds[objective_function]
    upper_bound = upper_bounds[objective_function]
    
    objective_function_name = objective_function_names[objective_function]
    ev.evolution(population_size, chromosome_size, generations, nr_breeding_parents, nr_offspring,
             nr_tournament, parent_selection_m, recomb_coef_m, objective_function, objective_function_name,
             lower_bound, upper_bound)
    